### PR TITLE
opt: Add support for index hints in UPDATE and DELETE statements

### LINKED ISSUE #38799: opt: support index hints on DELETE/UPDATE
Today, we silently ignore index hints on `UPDATE`/`DELETE`:

```
exec-ddl
CREATE TABLE x (
  a INT PRIMARY KEY,
  b INT,
  c INT,
  INDEX b_idx (b),
  INDEX c_idx (c)
)
----

opt
DELETE FROM
    x@primary
WHERE
    a = 1 AND b = 2 AND c = 3
----
delete x
 ├── columns: <none>
 ├── fetch columns: a:4(int) b:5(int) c:6(int)
 ├── cardinality: [0 - 0]
 ├── side-effects, mutations
 └── select
      ├── columns: a:4(int!null) b:5(int!null) c:6(int!null)
      ├── cardinality: [0 - 1]
      ├── key: ()
      ├── fd: ()-->(4-6)
      ├── index-join x
      │    ├── columns: a:4(int!null) b:5(int) c:6(int)
      │    ├── cardinality: [0 - 1]
      │    ├── key: ()
      │    ├── fd: ()-->(4-6)
      │    └── scan x@b_idx
      │         ├── columns: a:4(int!null) b:5(int!null)
      │         ├── constraint: /5/4: [/2/1 - /2/1]
      │         ├── cardinality: [0 - 1]
      │         ├── key: ()
      │         └── fd: ()-->(4,5)
      └── filters
           └── c = 3 [type=bool, outer=(6), constraints=(/6: [/3 - /3]; tight), fd=()-->(6)]
```

We should use the index requested by the hint or explicitly error out when one is provided (preferably the former)

### PR DESCRIPTION
**Pull request by savoie**
Client can now override index selection in UPDATE and DELETE statements (as was already possible with SELECT). While the index hints are not rejected in parsing, in query planning the index hint has previously been ignored.

An example of index hinting syntax that is now passed down into planning:

```
UPDATE t@primary SET ...
DELETE FROM t@idx ...
```

Fixes #38799

Related: #31012

Release justification: Low risk, high benefit changes to existing functionality.

### CODE REVIEW

**Review by justinj — COMMENTED** [review]
Nice, this looks great, and super fast! Would you mind also adding a test to the `execbuilder` tests? Take a look in `pkg/sql/opt/exec/execbuilder/testdata/{update,delete}`. I don't expect that there are any problems but it's good to make sure the flags get plumbed all the way through.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-:-Loag6B51FGMqhLKVKAi:b-t6mx8i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @savoie)

___
*[pkg/sql/opt/optbuilder/delete.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-LoagcuX07nt5xPJHb-d:-LoagcuX07nt5xPJHb-e:b-enzt45) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b74947472956c72d5c2a21f4fa886a384b3142/pkg/sql/opt/optbuilder/delete.go#L65)):*
> ```Go
> 
> 	var indexFlags *tree.IndexFlags
> 	if source, ok := del.Table.(*tree.AliasedTableExpr); ok {
> ```

nit: I think it would make sense to push these into `buildInputFor{Delete,Update}`, since those functions are already in the business of analyzing the syntax tree stuff relevant to the input.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-LoafTBADHYO2YESt8CA:-LoafTBB7XRMo4Mdk4FZ:bhtv6z3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b74947472956c72d5c2a21f4fa886a384b3142/pkg/sql/opt/optbuilder/mutation_builder.go#L212)):*
> ```Go
> func (mb *mutationBuilder) buildInputForUpdate(
> 	inScope *scope, from tree.TableExprs, where *tree.Where, limit *tree.Limit, orderBy tree.OrderBy,
> 	indexFlags *tree.IndexFlags,
> ```

nit: you should configure your editor to run [`crlfmt`](https://github.com/cockroachdb/crlfmt) automatically, not sure how that's done with emacs  (I can help with vim or Goland, though), maybe [this](https://wiki.crdb.io/wiki/spaces/CRDB/pages/73105658/Ben+s+Go+Emacs+Setup) or ask in #emacs. I think lint will reject this as-is
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 311 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-LoafnnDDkzDkabKYOOL:-LoafnnDDkzDkabKYOOM:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b74947472956c72d5c2a21f4fa886a384b3142/pkg/sql/opt/optbuilder/mutation_builder.go#L311)):*
> ```Go
> func (mb *mutationBuilder) buildInputForDelete(
> 	inScope *scope, where *tree.Where, limit *tree.Limit, orderBy tree.OrderBy,
> 	indexFlags *tree.IndexFlags,
> ```

ditto
___
*[pkg/sql/opt/optbuilder/update.go, line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-LoahKUw4BYA_6_XzKOp:-LoahKUw4BYA_6_XzKOq:b-qbkhft) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b74947472956c72d5c2a21f4fa886a384b3142/pkg/sql/opt/optbuilder/update.go#L100)):*
> ```Go
> 
> 	var indexFlags *tree.IndexFlags
> 	if source, ok := upd.Table.(*tree.AliasedTableExpr); ok {
> ```

ditto the comment in `buildDelete`
___
*[pkg/sql/opt/optbuilder/testdata/delete, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-LoaiUP83BhADxSQ059U:-LoaiUP83BhADxSQ059V:b9vasiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b74947472956c72d5c2a21f4fa886a384b3142/pkg/sql/opt/optbuilder/testdata/delete#L95)):*
> ```
>       └── const: 10 [type=int]
> 
> # DELETE with index hints
> ```

ditto the period
___
*[pkg/sql/opt/optbuilder/testdata/update, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-LoahPxH4KCoZ99sq_i2:-LoahPxI8yalaNCf-mD8:butqhvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b74947472956c72d5c2a21f4fa886a384b3142/pkg/sql/opt/optbuilder/testdata/update#L286)):*
> ```
>                 └── const: 1 [type=int]
> 
> # UPDATE with index hints
> ```

nit: comment guidelines are to have periods, we don't observe this perfectly, but we should
___
*[pkg/sql/opt/optbuilder/testdata/update, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-LoahOki8OkE0qwrNyNm:-LoahOki8OkE0qwrNyNn:b1w37ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b74947472956c72d5c2a21f4fa886a384b3142/pkg/sql/opt/optbuilder/testdata/update#L297)):*
> ```
> ----
> 
> build
> ```

nice tests! what happens if I refer to an index that doesn't exist?

**Review by RaduBerinde — COMMENTED** [review]
Impressive that you got this done already!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-:-LoakUMAAocAiK0gNi4s:bmrmqvl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @savoie)

**Review by rytaft — APPROVED** [review]
Nicely done! :lgtm:

> Solves #38799

[nit] If you change "Solves" to "Fixes" or "Closes" in the PR message, a bot will actually close the issue when this PR merges

[nit] I'd probably squash your commits into one before merging. Also, you should put the text from the PR message into the commit message as well. I think this also deserves a release note in the commit message explaining the change.

> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-:-LobUQwm7esXhfkUGggG:boqyw7b)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @savoie)

**Comment by RaduBerinde** [comment]
LGTM (but we'll need to wait for the release branching before we merge)

**Comment by RaduBerinde** [comment]
@andy-kimball: After talking to @awoods187 I think there is enough benefit for this fix to fit under "Low risk, high benefit changes to existing functionality". Let me know what you think.

**Review by andy-kimball — APPROVED** [review]
Agreed, this looks like a very small, safe change that will bring big benefits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-:-Lof1TasCheS5S0QFWxj:bw1q8eu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @savoie)

**Comment by RaduBerinde** [comment]
Awesome. @savoie can you squash and add the things below to the commit message?
```
Release note (bug fix): index hints used with DELETE and UPDATE are now taken into consideration as expected
Release justification: Low risk, high benefit changes to existing functionality.
```

**Comment by ericharmeling** [comment]
@awoods187 
Thanks for the heads-up.
We'll generate a docs issue from the release notes that @RaduBerinde just requested, so no need for an airtable entry.

**Comment by savoie** [comment]
Added the additional context to the commit message.

**Review by rytaft — APPROVED** [review]
New commit message :lgtm: 

Looks like you'll need to rebase from master to resolve the merge conflict

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-:-LouIWu6D8S5XIXthlSn:b-v4rsz9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @savoie)

**Review by justinj — COMMENTED** [review]
:lgtm: great job!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40715#-:-LouWtvRE4vuSMouUlK3:b-g547tm)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @savoie)

**Comment by RaduBerinde** [comment]
Looks like some execbuilder tests need updating - you can run with the `-rewrite` flag and it does that for you: `make test PKG=./pkg/sql/opt/exec/execbuilder TESTFLAGS=-rewrite`

**Pull request merged**
